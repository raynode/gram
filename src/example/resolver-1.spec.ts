import {
  graphql,
  GraphQLBoolean,
  GraphQLInt,
  GraphQLString,
  printSchema,
} from 'graphql'
import { createSchemaBuilder, Service } from '..'

interface Animal {
  type: string
  name: string
  tame: boolean
  age: number
}

interface DatabaseAnimal {
  type: string
  name: string
  tame: boolean
  birthdate: number
}

let animals: any[] = []

const Animals: Service<Animal, DatabaseAnimal> = {
  findMany: async () => ({
    page: { limit: 100, offset: 0 },
    nodes: animals,
  }),
  findOne: async () => null,
  create: async () => null,
  remove: async () => null,
  update: async () => null,
}

describe('testing the example 1', () => {
  it('should build the example code', () => {
    const builder = createSchemaBuilder()
    const animal = builder.model('Animal', Animals)
    // animal type like 'dog', 'cat'
    // field is required
    animal.attr('type', GraphQLString).isNotNullable()
    // animal name like 'Fluffy', 'Rex'
    // field is not required
    animal.attr('name', GraphQLString)
    // is it a tame animal
    animal.attr('tame', GraphQLBoolean)
    // age of the animal
    animal.attr('age', GraphQLInt)

    animal.resolve(() => ({
      age: animal =>
        Math.floor((Date.now() - animal.birthdate) / 1000 / 60 / 60 / 24 / 365),
    }))

    expect(printSchema(builder.build())).toMatchSnapshot()
  })

  it('should use the resolver when queried', async () => {
    // initialize the database
    const birthdate = new Date()
    // create random age
    const age = Math.floor(Math.random() * 5 + 5)
    // change the year to be that amount of years back
    birthdate.setUTCFullYear(birthdate.getUTCFullYear() - age)
    animals = [
      {
        type: 'cat',
        name: 'Fluffy',
        tame: true,
        birthdate: +birthdate, // convert to number
      },
    ]
    const builder = createSchemaBuilder()
    const animal = builder.model('Animal', Animals)
    // animal type like 'dog', 'cat'
    // field is required
    animal.attr('type', GraphQLString).isNotNullable()
    // animal name like 'Fluffy', 'Rex'
    // field is not required
    animal.attr('name', GraphQLString)
    // is it a tame animal
    animal.attr('tame', GraphQLBoolean)
    // age of the animal
    animal.attr('age', GraphQLInt)

    let called = false
    animal.resolve(() => ({
      age: animal => {
        called = true
        return Math.floor(
          (Date.now() - animal.birthdate) / 1000 / 60 / 60 / 24 / 365,
        )
      },
    }))

    const { resolvers } = builder.createBuild().toTypeDefs()
    const schema = builder.build()

    const { data, errors } = await graphql({
      schema,
      source: `{
        getAnimals(where: {}) {
          nodes {
            type
            name
            tame
            age
          }
        }
      }`,
    })
    // expect the age to be calculated
    expect(data.getAnimals.nodes[0].age).toEqual(age)
    // make sure the default resolvers are still active
    expect(data.getAnimals.nodes[0].name).toEqual('Fluffy')
    // expect the function to have been called
    expect(called).toBeTruthy()
  })
})
