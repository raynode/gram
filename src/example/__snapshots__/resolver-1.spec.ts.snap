// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the example 1 should build the example code 1`] = `
"type Animal implements Node {
  type: String!
  name: String
  tame: Boolean
  age: Int
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

input AnimalFilter {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  type: String
  type_not: String
  type_in: [String!]
  type_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input AnimalPage {
  limit: Int
  offset: Int
}

type Animals implements List {
  page: Page
  nodes: [Animal!]!
}

enum AnimalSortOrder {
  type_ASC
  type_DESC
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
}

input AnimalWhere {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  type: String
  type_not: String
  type_in: [String!]
  type_not_in: [String!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
}

input CreateAnimalData {
  type: String!
  name: String
  tame: Boolean
  age: Int
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

scalar DateTime

interface List {
  page: Page
  nodes: [Node!]!
}

type Mutation {
  createAnimal(data: CreateAnimalData!): Animal
  updateAnimal(data: UpdateAnimalData!, where: AnimalWhere!): [Animal!]!
  deleteAnimals(where: AnimalWhere!): [Animal!]!
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Page {
  page: Int
  limit: Int
  offset: Int
}

type Query {
  getAnimal(where: AnimalWhere!, order: AnimalSortOrder): Animal
  getAnimals(order: AnimalSortOrder, page: AnimalPage, where: AnimalWhere!): Animals
}

type Subscription {
  onCreateAnimal: Animal!
  onUpdateAnimal: [Animal!]!
  onDeleteAnimals: [Animal!]!
}

input UpdateAnimalData {
  type: String!
  name: String
  tame: Boolean
  age: Int
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}
"
`;
